sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/errorhandling/BadRecordRouter.java

```diff
@@ -0,0 +1,140 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.errorhandling;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.io.Serializable;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.transforms.DoFn.MultiOutputReceiver;
+import org.apache.beam.sdk.transforms.errorhandling.BadRecord.Failure;
+import org.apache.beam.sdk.transforms.errorhandling.BadRecord.Record;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.util.Preconditions;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Charsets;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public interface BadRecordRouter extends Serializable {
+
+  BadRecordRouter THROWING_ROUTER = new ThrowingBadRecordRouter();
+
+  BadRecordRouter RECORDING_ROUTER = new RecordingBadRecordRouter();
+
+  TupleTag<BadRecord> BAD_RECORD_TAG = new TupleTag<>();
+
+  <RecordT> void route(
+      MultiOutputReceiver outputReceiver,
+      RecordT record,
+      @Nullable Coder<RecordT> coder,
+      @Nullable Exception exception,
+      String description,
+      String failingTransform)
+      throws Exception;
+
+  class ThrowingBadRecordRouter implements BadRecordRouter {
+
+    @Override
+    public <RecordT> void route(
+        MultiOutputReceiver outputReceiver,
+        RecordT record,
+        @Nullable Coder<RecordT> coder,
+        @Nullable Exception exception,
+        String description,
+        String failingTransform)
+        throws Exception {
+      if (exception != null) {
+        throw exception;
+      }
+    }
+  }
+
+  class RecordingBadRecordRouter implements BadRecordRouter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RecordingBadRecordRouter.class);
+
+    @Override
+    public <RecordT> void route(
+        MultiOutputReceiver outputReceiver,
+        RecordT record,
+        @Nullable Coder<RecordT> coder,
+        @Nullable Exception exception,
+        String description,
+        String failingTransform)
+        throws Exception {
+      Preconditions.checkArgumentNotNull(record);
+      ObjectWriter objectWriter = new ObjectMapper().writer().withDefaultPrettyPrinter();
+
+      // Build up record information
```

Comment:
I wonder if this logic would be better placed as factory methods on BadRecord itself.

Reply:
I was thinking the whole block down to line 92.